# **Ejercicios de pair programming Módulo 3 Sprint 1**
## **Regreción Lineal: Lección 7 - ANOVA**

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.


Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.


📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [13]:

import numpy as np
import pandas as pd
import random 

import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

In [14]:
df = pd.read_csv("datos/soci_econ_country_profiles_estandarizado.csv", index_col = 0)
df.sample(5)

country             region  restaurant_price_index  population_2017  \
40  Philippines  South-easternAsia                   19.86         1.465284   
1     Australia            Oceania                   77.12         0.095085   
39     Pakistan       SouthernAsia                   21.01         3.033537   
13      Croatia     SouthernEurope                   37.53        -0.249938   
65      Vietnam  South-easternAsia                   20.46         1.305612   

    employment_agriculture  employment_industry  employment_services  \
40                1.797101            -1.048611            -0.659971   
1                -0.273292            -0.368056             0.487805   
39                2.989648            -0.562500            -1.692970   
13                0.265010             0.409722            -0.206600   
65                2.964803            -0.131944            -1.859397   

    education  quality_of_life_index  purchasing_power_index  safety_index  \
40  -3.395530              -1.306713               -0.708552      0.007199   
1    0.211765               0.550795                0.660990     -0.139554   
39  -1.011765              -0.739542               -0.626512     -0.801883   
13  -0.070588               0.459061               -0.195337      0.670082   
65   0.447059              -1.628793               -0.941793     -0.723245   

    cost_of_living  consumer_price_index  human_capital_index  inflation  \
40       -0.436772              0.326423            -0.946588   0.383678   
1         0.893666              0.102195             0.566766  -0.030665   
39       -0.654796              2.145194            -1.890208   0.948106   
13       -0.051972             -0.310264             0.091988  -0.405950   
65       -0.280420              1.982659            -0.246291   0.689243   

    urban_population  population_age_distribution_0_14  \
40         -1.029412                          1.720117   
1           0.393466                          0.239067   
39         -1.400894                          2.081633   
13         -0.667181                         -0.262391   
65         -1.445479                          0.717201   

    population_age_distribution_60_plus  international_migrant_stock  \
40                            -1.026415                    -0.670611   
1                             -0.015094                     1.538462   
39                            -1.094340                    -0.536489   
13                             0.422642                     0.386588   
65                            -0.762264                    -0.678501   

    restaurant_price_index_normal  
40                       3.176574  
1                        4.749998  
39                       3.240159  
13                       3.903968  
65                       3.210177

In [15]:
df.columns

Index(['country', 'region', 'restaurant_price_index', 'population_2017',
       'employment_agriculture', 'employment_industry', 'employment_services',
       'education', 'quality_of_life_index', 'purchasing_power_index',
       'safety_index', 'cost_of_living', 'consumer_price_index',
       'human_capital_index', 'inflation', 'urban_population',
       'population_age_distribution_0_14',
       'population_age_distribution_60_plus', 'international_migrant_stock',
       'restaurant_price_index_normal'],
      dtype='object')

In [16]:
# iniciamos nuestro anova. Recrestaurant_price_index_normalordemos que la variable respuesta es el precio de las casas (median_house_value)
lm = ols('''restaurant_price_index_normal ~ region + population_2017 + employment_agriculture +
       employment_industry + employment_services + education +
       quality_of_life_index + purchasing_power_index + safety_index +
       cost_of_living + consumer_price_index + human_capital_index +
       inflation + urban_population + population_age_distribution_0_14 +
       population_age_distribution_60_plus + international_migrant_stock''', 
         data=df).fit()
sm.stats.anova_lm(lm)

df     sum_sq   mean_sq           F  \
region                               13.0  15.935057  1.225774   46.662113   
population_2017                       1.0   0.437499  0.437499   16.654467   
employment_agriculture                1.0   2.649718  2.649718  100.868095   
employment_industry                   1.0   0.235514  0.235514    8.965425   
employment_services                   1.0   0.253964  0.253964    9.667764   
education                             1.0   0.004721  0.004721    0.179726   
quality_of_life_index                 1.0   0.009399  0.009399    0.357782   
purchasing_power_index                1.0   0.057461  0.057461    2.187399   
safety_index                          1.0   0.020066  0.020066    0.763877   
cost_of_living                        1.0   2.797179  2.797179  106.481549   
consumer_price_index                  1.0   0.010235  0.010235    0.389632   
human_capital_index                   1.0   0.043117  0.043117    1.641358   
inflation                             1.0   0.156283  0.156283    5.949305   
urban_population                      1.0   0.171459  0.171459    6.527005   
population_age_distribution_0_14      1.0   0.000239  0.000239    0.009087   
population_age_distribution_60_plus   1.0   0.071716  0.071716    2.730056   
international_migrant_stock           1.0   0.012273  0.012273    0.467195   
Residual                             36.0   0.945689  0.026269         NaN   

                                           PR(>F)  
region                               1.496188e-18  
population_2017                      2.378539e-04  
employment_agriculture               5.534266e-12  
employment_industry                  4.951272e-03  
employment_services                  3.655035e-03  
education                            6.741312e-01  
quality_of_life_index                5.534854e-01  
purchasing_power_index               1.478418e-01  
safety_index                         3.879111e-01  
cost_of_living                       2.666975e-12  
consumer_price_index                 5.364290e-01  
human_capital_index                  2.083308e-01  
inflation                            1.978250e-02  
urban_population                     1.499748e-02  
population_age_distribution_0_14     9.245854e-01  
population_age_distribution_60_plus  1.071728e-01  
international_migrant_stock          4.986566e-01  
Residual                                      NaN

In [17]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                  OLS Regression Results                                 
=========================================================================================
Dep. Variable:     restaurant_price_index_normal   R-squared:                       0.960
Model:                                       OLS   Adj. R-squared:                  0.928
Method:                            Least Squares   F-statistic:                     30.02
Date:                           Thu, 11 May 2023   Prob (F-statistic):           1.13e-17
Time:                                   22:16:33   Log-Likelihood:                 46.451
No. Observations:                             66   AIC:                            -32.90
Df Residuals:                                 36   BIC:                             32.79
Df Model:                                     29                                         
Covariance Type:                       nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               3.1332      1.319      2.375      0.023       0.458       5.809
region[T.EasternAsia]                  -0.3768      0.253     -1.489      0.145      -0.890       0.136
region[T.EasternEurope]                 0.0808      0.222      0.364      0.718      -0.370       0.531
region[T.NorthernAfrica]                0.8828      0.395      2.234      0.032       0.081       1.684
region[T.NorthernAmerica]               0.1873      0.251      0.747      0.460      -0.321       0.696
region[T.NorthernEurope]                0.3170      0.223      1.423      0.163      -0.135       0.769
region[T.Oceania]                       0.0779      0.244      0.320      0.751      -0.416       0.572
region[T.South-easternAsia]            -0.4003      0.234     -1.714      0.095      -0.874       0.073
region[T.SouthAmerica]                 -0.0193      0.199     -0.097      0.923      -0.423       0.384
region[T.SouthernAfrica]                0.7143      0.313      2.284      0.028       0.080       1.349
region[T.SouthernAsia]                  0.1426      0.254      0.561      0.579      -0.373       0.658
region[T.SouthernEurope]                0.0814      0.234      0.348      0.730      -0.393       0.556
region[T.WesternAsia]                   0.1749      0.201      0.869      0.391      -0.234       0.583
region[T.WesternEurope]                 0.2436      0.239      1.021      0.314      -0.240       0.727
population_2017                         0.0110      0.008      1.398      0.171      -0.005       0.027
employment_agriculture                  3.8875      6.418      0.606      0.549      -9.129      16.904
employment_industry                     2.3022      3.818      0.603      0.550      -5.440      10.045
employment_services                     5.6169      9.250      0.607      0.548     -13.144      24.377
education                               0.0012      0.002      0.492      0.626      -0.004       0.006
quality_of_life_index                   0.0773      0.108      0.719      0.477      -0.141       0.295
purchasing_power_index                 -0.2593      0.098     -2.645      0.012      -0.458      -0.060
safety_index                           -0.0626      0.061     -1.031      0.310      -0.186       0.061
cost_of_living                          0.6338      0.078      8.082      0.000       0.475       0.793
consumer_price_index                    0.0007      0.027      0.025      0.980      -0.054       0.055
human_capital_index                     0.2686      0.105      2.547      0.015       0.055       0.482
inflation                              -0.0758    

- El R-Square de nuetros datos es muy alto, llegando a 0.96, es decir, los datos sugieren que nuestro modelo de predición es muy fiable porque explica el 96% de la variación de la variable respuesta. Sin embargo, estas conclusiones son fictícias ya que nuestros datos no siguen una distribución normal. 
  
- Las variables que afectan la variable respuesta (pvalue <= 0.05) en nuetros datos son: 'region', 'purchasing_power_index', 'cost_of_living', 'human_capital_index', 'inflation' y 'urban_population'. 
  
- De las columnas que afectan nuestra VR sacamos los coef: 'region' (con distintos valores), 'purchasing_power_index' (coef: -0.2593), 'cost_of_living' (coef 0.6338), 'human_capital_index' (coef 0.2686), 'inflation' (coef -0.0758) y 'urban_population' (coef 0.2021). En el caso de la variable 'purchasing_power_index' (coef: -0.2593), por cada unidad que aumenta 'purchasing_power_index', la variable respuesta disminuye 0.2593. 